File: packages/sdk-ts/src/pagination.ts
// SPDX-License-Identifier: Apache-2.0

// Auto-iterating pagination helpers — the TypeScript mirror of
// `af_stack/pagination.py` (kept 1:1 for cross-SDK parity).
//
// List endpoints across the suite paginate two ways: offset based
// (`limit`/`offset` with `total`/`has_more`) and cursor based (`next_token`).
// `AsyncPaginator` unifies both behind a single async-iterable so callers
// never hand-roll a fetch loop:
//
//   import { BackAI, paginate } from "@af-stack/sdk"
//
//   const client = new BackAI()
//   const pager = paginate<Job>(async (cursor) => {
//     const offset = (cursor as number) ?? 0
//     const page = await client.jobs.list({ offset, limit: 50 })
//     const next = page.hasMore ? offset + page.jobs.length : null
//     return [page.jobs, next]
//   })
//   for await (const job of pager) console.log(job.id)
//
// The page fetcher is deliberately transport-agnostic: you provide the page
// fetch, it walks the cursors.

/**
 * A page fetcher takes the current cursor (`null` for the first page) and
 * returns `[items, nextCursor]`. A `nextCursor` of `null` ends iteration.
 */
export type PageFetcher<T> = (cursor: unknown) => Promise<[T[], unknown]> | [T[], unknown]

/** Wrap a page fetcher into an async-iterable over individual items. */
export class AsyncPaginator<T> implements AsyncIterable<T> {
  private readonly fetchPage: PageFetcher<T>
  private readonly startCursor: unknown

  constructor(fetchPage: PageFetcher<T>, startCursor: unknown = null) {
    this.fetchPage = fetchPage
    this.startCursor = startCursor
  }

  async *[Symbol.asyncIterator](): AsyncIterator<T> {
    let cursor = this.startCursor
    while (true) {
      const [items, nextCursor] = await this.fetchPage(cursor)
      for (const item of items) yield item
      if (nextCursor === null || nextCursor === undefined) break
      cursor = nextCursor
    }
  }

  /** Iterate page-by-page instead of item-by-item. */
  async *pages(): AsyncIterable<T[]> {
    let cursor = this.startCursor
    while (true) {
      const [items, nextCursor] = await this.fetchPage(cursor)
      yield items
      if (nextCursor === null || nextCursor === undefined) break
      cursor = nextCursor
    }
  }

  /** Eagerly drain the paginator into an array (optionally capped). */
  async collect(opts: { limit?: number } = {}): Promise<T[]> {
    const out: T[] = []
    for await (const item of this) {
      out.push(item)
      if (opts.limit !== undefined && out.length >= opts.limit) break
    }
    return out
  }
}

/**
 * Return an {@link AsyncPaginator} over `fetchPage`. `fetchPage(cursor)` must
 * return `[items, nextCursor]`; a `nextCursor` of `null` stops iteration.
 */
export function paginate<T>(
  fetchPage: PageFetcher<T>,
  opts: { startCursor?: unknown } = {},
): AsyncPaginator<T> {
  return new AsyncPaginator<T>(fetchPage, opts.startCursor ?? null)
}

File: packages/sdk-py/af_stack/pagination.py
# SPDX-License-Identifier: Apache-2.0

"""Auto-iterating pagination helpers.

List endpoints across the suite paginate two ways: **offset** based
(``limit`` / ``offset`` with ``total`` / ``has_more``) and **cursor** based
(``next_token``). :class:`AsyncPaginator` unifies both behind a single
async-iterable so callers never hand-roll a fetch loop::

    from af_stack import BackAI, paginate

    client = BackAI()

    async def _page(cursor):
        result = await client.jobs.list(offset=cursor or 0, limit=50)
        next_cursor = (cursor or 0) + len(result.jobs) if result.has_more else None
        return result.jobs, next_cursor

    async for job in paginate(_page):
        print(job.id)

The :func:`paginate` helper is deliberately transport-agnostic — you provide
the page fetch, it walks the cursors. This mirrors the TypeScript
``paginate`` helper 1:1 for cross-SDK parity.
"""

from __future__ import annotations

from collections.abc import AsyncIterator, Awaitable, Callable
from typing import Generic, TypeVar

T = TypeVar("T")
Cursor = object

# A page fetcher takes the current cursor (``None`` for the first page) and
# returns ``(items, next_cursor)``. A ``next_cursor`` of ``None`` ends
# iteration.
PageFetcher = Callable[[object | None], Awaitable["tuple[list[T], object | None]"]]


class AsyncPaginator(Generic[T]):
    """Wrap a page fetcher into an async-iterable over individual items."""

    def __init__(
        self,
        fetch_page: PageFetcher[T],
        *,
        start_cursor: object | None = None,
    ) -> None:
        self._fetch_page = fetch_page
        self._start_cursor = start_cursor

    def __aiter__(self) -> AsyncIterator[T]:
        return self._iterate()

    async def _iterate(self) -> AsyncIterator[T]:
        cursor = self._start_cursor
        while True:
            items, next_cursor = await self._fetch_page(cursor)
            for item in items:
                yield item
            if next_cursor is None:
                break
            cursor = next_cursor

    async def pages(self) -> AsyncIterator[list[T]]:
        """Iterate page-by-page instead of item-by-item."""
        cursor = self._start_cursor
        while True:
            items, next_cursor = await self._fetch_page(cursor)
            yield items
            if next_cursor is None:
                break
            cursor = next_cursor

    async def collect(self, *, limit: int | None = None) -> list[T]:
        """Eagerly drain the paginator into a list (optionally capped)."""
        out: list[T] = []
        async for item in self:
            out.append(item)
            if limit is not None and len(out) >= limit:
                break
        return out


def paginate(
    fetch_page: PageFetcher[T],
    *,
    start_cursor: object | None = None,
) -> AsyncPaginator[T]:
    """Return an :class:`AsyncPaginator` over ``fetch_page``.

    ``fetch_page(cursor)`` must return ``(items, next_cursor)``; a
    ``next_cursor`` of ``None`` stops iteration.
    """
    return AsyncPaginator(fetch_page, start_cursor=start_cursor)


__all__ = ["AsyncPaginator", "paginate"]
